### PR TITLE
networkctl: Add support to display bridge properties.

### PR DESCRIPTION
**Pull request by ssahani**
How do we arrange this . Bridge has lots of data.

```
● 5: virbr0                                                     
             Link File: /usr/lib/systemd/network/99-default.link
          Network File: n/a                                     
                  Type: bridge                                  
                 State: no-carrier (unmanaged)                  
                Driver: bridge                                  
            HW Address: 52:54:00:d3:e4:80                       
                   MTU: 1500 (min: 68, max: 65535)              
         forward_delay: 200                                     
            hello_time: 200                                     
               max_age: 2000                                    
           ageing_time: 30000                                   
              priority: 32768                                   
             stp_state: 1                                       
    mcast_igmp_version: 2                                       
  Queue Length (Tx/Rx): 1/1                                     
               Address: 192.168.122.1 

```

### CODE REVIEW

File: src/network/networkctl.c [hunk 1]
```diff
@@ -202,6 +213,32 @@ static int decode_link(sd_netlink_message *m, LinkInfo *info, char **patterns) {
...
 
+        /* fill kind info */
+        r = sd_netlink_message_enter_container(m, IFLA_LINKINFO);
+        if (r >= 0) {
+                r = sd_netlink_message_read_string(m, IFLA_INFO_KIND, &received_kind);
+                if (r >= 0) {
+                        r = sd_netlink_message_enter_container(m, IFLA_INFO_DATA);
+                        if (r >= 0) {
+                                if (streq(received_kind, "bridge")) {
+                                        (void) sd_netlink_message_read_u32(m, IFLA_BR_FORWARD_DELAY, &info->forward_delay);
```
Comment (yuwata) [inline src/network/networkctl.c]: Horrible indentation :-) It is better to move them to, say, `decode_bridge_attrs()`.

File: src/network/networkctl.c [hunk 2]
```diff
@@ -1130,6 +1167,34 @@ static int link_status_one(
...
+
+                        xsprintf(forward_delay_str, "%" PRIu8, info->forward_delay);
+
+                        r = table_add_cell(table, NULL, TABLE_EMPTY, NULL);
+                        if (r < 0)
+                                return r;
+                        r = table_add_cell(table, NULL, TABLE_STRING, "forward_delay:");
+                        if (r < 0)
+                                return r;
+                        r = table_add_cell(table, NULL, TABLE_STRING, forward_delay_str);
```
Comment (yuwata) [inline src/network/networkctl.c]: Please use `table_add_cell_stringf()`.

File: src/network/networkctl.c [hunk 3]
```diff
@@ -123,6 +123,17 @@ typedef struct LinkInfo {
         uint64_t tx_bitrate;
         uint64_t rx_bitrate;
 
+        char *netdev_kind;
```
Comment (yuwata) [inline src/network/networkctl.c]: This is not `free()`d.

**Comment by Nadyuwka** [comment]
ср, 24 июл 2019, 19:56 Susant Sahani <notifications@github.com>:

File: src/network/networkctl.c (hunk 1 above)
Comment (ssahani) [reply src/network/networkctl.c]: Hahaha Well I am having fun with this will make a separate func and add them all. thanks

**Comment by ssahani** [comment]
updated PTAL @yuwata .

**Comment by ssahani** [comment]
Also Tunnel/VXlan work is in WIP.

File: src/network/networkctl.c [hunk 4]
```diff
@@ -142,10 +152,47 @@ static int link_info_compare(const LinkInfo *a, const LinkInfo *b) {
...
+        (void) sd_netlink_message_read_u32(m, IFLA_BR_MAX_AGE, &info->max_age);
+        (void) sd_netlink_message_read_u32(m, IFLA_BR_AGEING_TIME, &info->ageing_time);
+        (void) sd_netlink_message_read_u32(m, IFLA_BR_STP_STATE, &info->stp_state);
+        (void) sd_netlink_message_read_u16(m, IFLA_BR_PRIORITY, &info->priority);
+        (void) sd_netlink_message_read_u8(m, IFLA_BR_MCAST_IGMP_VERSION, &info->mcast_igmp_version);
+
+        strncpy(info->netdev_kind, received_kind, IFNAMSIZ);
+
+        (void) sd_netlink_message_exit_container(m);
+
```
Comment (yuwata) [inline src/network/networkctl.c:189]: One more `exit_container()` for `IFLA_LINKINFO`?

File: src/network/networkctl.c [hunk 5]
```diff
@@ -106,6 +106,7 @@ static void setup_state_to_color(const char *state, const char **on, const char
 
 typedef struct LinkInfo {
         char name[IFNAMSIZ+1];
+        char netdev_kind[IFNAMSIZ+1];
```
Comment (yuwata) [inline src/network/networkctl.c]: Could you tell me where the maximum length of `kind` is defined in kernel?

File: src/network/networkctl.c [hunk 6]
```diff
@@ -1130,6 +1180,91 @@ static int link_status_one(
                         return r;
         }
 
+        if (streq_ptr(info->netdev_kind, "bridge")) {
+                if (info->forward_delay) {
+                        r = table_add_cell(table, NULL, TABLE_EMPTY, NULL);
+                        if (r < 0)
+                                return r;
+                        r = table_add_cell(table, NULL, TABLE_STRING, "forward_delay:");
```
Comment (yuwata) [inline src/network/networkctl.c]: Please capitalize the field name: `Forward Delay:`, `Hello Time:`, or `Max Age:`.

File: src/network/networkctl.c [hunk 7]
```diff
@@ -1130,6 +1180,91 @@ static int link_status_one(
                         return r;
         }
 
+        if (streq_ptr(info->netdev_kind, "bridge")) {
+                if (info->forward_delay) {
```
Comment (yuwata) [inline src/network/networkctl.c:1185]: `if (info->forward_delay > 0)`

File: src/network/networkctl.c (hunk 5 above)
Comment (ssahani) [reply src/network/networkctl.c]: it's defined in one place like ```MODULE_NAME_LEN 64 - sizeof(Elf_Addr))``` so I think 54 should be OK.

File: src/network/networkctl.c (hunk 7 above)
Comment (ssahani) [reply src/network/networkctl.c:1185]: We can delay to 0. ```sudo ip link set br1 type bridge forward_delay 0```

**Comment by ssahani** [comment]
updated thanks for review .

**Pull request closed without merging**